core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.kt

```diff
@@ -0,0 +1,677 @@
+     */
+    fun sliceIndex(beginStepIndex: Int, endStepIndex: Int): EnvelopePart? {
+        assert(endStepIndex >= 0 && endStepIndex <= stepCount())
+        assert(beginStepIndex >= 0 && beginStepIndex <= stepCount())
+        assert(beginStepIndex <= endStepIndex)
+
+        val resultSize = endStepIndex - beginStepIndex
+        if (resultSize <= 0) return null
+
+        val slicePos = Arrays.copyOfRange(positions, beginStepIndex, endStepIndex + 1)
+        val sliceSpeeds = Arrays.copyOfRange(speeds, beginStepIndex, endStepIndex + 1)
+        val sliceTimes = Arrays.copyOfRange(timeDeltas, beginStepIndex, endStepIndex)
+        return EnvelopePart(attrs, slicePos, sliceSpeeds, sliceTimes)
+    }
+
+    fun sliceBeginning(endIndex: Int, endPosition: Double, endSpeed: Double): EnvelopePart? {
+        return slice(0, Double.NEGATIVE_INFINITY, Double.NaN, endIndex, endPosition, endSpeed)
+    }
+
+    fun sliceEnd(beginIndex: Int, beginPosition: Double, beginSpeed: Double): EnvelopePart? {
+        return slice(
+            beginIndex,
+            beginPosition,
+            beginSpeed,
+            stepCount() - 1,
+            Double.POSITIVE_INFINITY,
+            Double.NaN
+        )
+    }
+
+    /**
+     * Cuts an envelope part with imposed speeds on the edges
+     *
+     * @return an EnvelopePart spanning from beginPosition to endPosition
+     */
+    fun sliceWithSpeeds(
+        beginPosition: Double,
+        beginSpeed: Double,
+        endPosition: Double,
+        endSpeed: Double
+    ): EnvelopePart? {
+        var newBeginPos = beginPosition
+        var newEndPos = endPosition
+        var beginIndex = 0
+        if (newBeginPos <= beginPos) newBeginPos = Double.NEGATIVE_INFINITY
+        if (newBeginPos != Double.NEGATIVE_INFINITY) beginIndex = findRight(newBeginPos)
+        var endIndex = stepCount() - 1
+        if (newEndPos >= endPos) newEndPos = Double.POSITIVE_INFINITY
+        if (newEndPos != Double.POSITIVE_INFINITY) endIndex = findLeft(newEndPos)
+        return slice(beginIndex, newBeginPos, beginSpeed, endIndex, newEndPos, endSpeed)
+    }
+
+    /**
+     * Cuts an envelope part, interpolating new points if required.
+     *
+     * @return an EnvelopePart spanning from beginPosition to endPosition
+     */
+    fun slice(beginPosition: Double, endPosition: Double): EnvelopePart? {
+        var newBeginPos = beginPosition
+        var newEndPos = endPosition
+        var beginIndex = 0
+        if (newBeginPos <= beginPos) newBeginPos = Double.NEGATIVE_INFINITY
+        if (newBeginPos != Double.NEGATIVE_INFINITY) beginIndex = findRight(newBeginPos)
+        var endIndex = stepCount() - 1
+        if (newEndPos >= endPos) newEndPos = Double.POSITIVE_INFINITY
+        if (newEndPos != Double.POSITIVE_INFINITY) endIndex = findLeft(newEndPos)
+        return slice(beginIndex, newBeginPos, endIndex, newEndPos)
+    }
```

Comment:
Could really use default values to avoid that duplication, but the parameter ordering isn't the most convenient for an easy change